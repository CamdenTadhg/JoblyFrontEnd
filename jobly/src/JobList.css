:root{
    background-color: #D2D0BA;
    height: 100vh;
}