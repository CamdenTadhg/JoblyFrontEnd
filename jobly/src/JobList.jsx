import React from 'react';
import SearchBar from './SearchBar';
import JobCard from './JobCard';

function JobList(){
    return(
        <div>
            I am the job list page.
            <SearchBar />
            <JobCard />
        </div>
    )
}

export default JobList;