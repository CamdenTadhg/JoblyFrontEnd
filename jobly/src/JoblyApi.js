import axios from "axios";

// const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";
const BASE_URL = 'http://localhost:3001';

/** API Class.
 *
 * Static class tying together methods used to get/send to to the API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't
 * be any API-aware stuff elsewhere in the frontend.
 *
 */

class JoblyApi {
  // the token for interaction with the API will be stored here.
  static token;

  static async request(endpoint, data = {}, method = "get") {
    console.log("API Call:", endpoint, data, method);

    //there are multiple ways to pass an authorization token, this is how you pass it in the header.
    //this has been provided to show you another way to pass the token. you are only expected to read this code for this project.
    const url = `${BASE_URL}/${endpoint}`;
    const headers = { authorization: `Bearer ${JoblyApi.token}` };
    const params = (method === "get")
        ? data
        : {};

    try {
      console.log('headers are', headers);
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.response);
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    }
  }

  // Individual API routes
  /** Login a user and receive back a token. LoginData = {username, password}*/
  static async login(loginData){
    try {
      let res = await this.request(`auth/token`, loginData, 'post');
      console.log('res.token is ', res.token);
      JoblyApi.token=res.token;
      return res.token;
    } catch (error){
      return error;
    }
  }

  /** Signup a user and receive back a token. signupData = {username, password, firstName, lastName, email} */
  static async signup(signupData){
    try{
      let res = await this.request(`auth/register`, signupData, "post");
      JoblyApi.token = res.token;
      return res.token;
    } catch (error){
      return error;
    }
  }

  /** Get list of all companies */
  static async getAllCompanies(){
    let res = await this.request(`companies`);
    return res.companies;
  }

  /** Get a filtered list of companies. searchFilters = {minEmployees, maxEmployees, name} */
  static async getFilteredCompanies(searchFilters) {
    let {minEmployees, maxEmployees, name} = searchFilters;
    let url = 'companies?';
    if (minEmployees) {
      url = url + `minEmployees=${minEmployees}&`;
    }
    if (maxEmployees){
      url = url + `maxEmployees=${maxEmployees}&`;
    }
    if (name) {
      url = url + `name=${name}`
    }
    let res = await this.request(url);
    console.log(res);
    return res.companies;
  }

  /** Get details on a company by handle. */
  static async getCompany(handle) {
    let res = await this.request(`companies/${handle}`);
    console.log(res.company);
    return res.company;
  }

  /**Get list of all jobs */
  static async getAllJobs(){
    let res = await this.request(`jobs`)
    return res.jobs;
  }

  /** Get a filtered list of jobs. searchFilters = {salary, equity, title}*/
  static async getFilteredJobs(searchFilters){
    let {salary, equity, title} = searchFilters;
    let url = 'jobs?'
    if (salary){
      url = url + `minSalary=${salary}&`;
    }
    if (equity) {
      url=url+`hasEquity=${equity}&`;
    }
    if(title) {
      url= url + `title=${title}`;
    }
    let res = await this.request(url);
    return res.jobs;
  }

  /** Get a user's details */
  static async getUserDetails(username){
    let res = await this.request(`users/${username}`);
    return res.user;
  }

  /** Edit a user's details. userData = {firstName, lastName, password, email} */
  static async editUserDetails(username, userData){
    let res = await this.request(`users/${username}`, userData, 'patch');
    console.log('res is', res);
    return res.user;
  }

  /** Delete a user's account */
  static async deleteUser(username) {
    let res = await this.request(`users/${username}`, 'delete');
    return res.deleted;
  }

  /**Create job application for a user */
  static async applyToJob(username, jobId){
    let res = await this.request(`users/${username}/jobs/${jobId}`, {}, "post");
    return res;
  }
}

export default JoblyApi;
