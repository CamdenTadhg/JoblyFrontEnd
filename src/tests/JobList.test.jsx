import {render, waitFor, cleanup} from '@testing-library/react';
import {test, expect, vi, afterEach, beforeEach} from 'vitest';
import {MemoryRouter} from 'react-router-dom';
import JobList from '../JobList';
import JoblyApi from '../JoblyApi';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import UserContext from '../contexts/userContext'

const Wrapper = ({children}) => {
    return (
        <MemoryRouter>
            <UserContext.Provider value='testuser'>
                {children}
            </UserContext.Provider>
        </MemoryRouter>
    );
};

let spyGetAllJobs;

vi.mock('react-router-dom', async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...actual,
        Navigate: vi.fn(({to}) => (`Redirected to ${to}`))
    }
});

beforeEach(() => {
    const mock = new MockAdapter(axios);
    mock.onGet('http://localhost:3001/jobs').reply(200, {jobs: [
        {title: 'Accommodation manager', salary: 126000, equity: '0', companyHandle:'anderson-arias-morrow', applied: true},
        {title: 'Accountant, chartered certified', salary: 175000, equity: '0.1', companyHandle: 'arnold-berger-townsend', applied: false}
      ]});
    mock.onGet('http://localhost:3001/jobs?').reply(200, {jobs: [
        {title: 'Accommodation manager', salary: 126000, equity: '0', companyHandle:'anderson-arias-morrow', applied: true},
        {title: 'Accountant, chartered certified', salary: 175000, equity: '0.1', companyHandle: 'arnold-berger-townsend', applied: false}
      ]})
    spyGetAllJobs = vi.spyOn(JoblyApi, 'getAllJobs');
});

afterEach(cleanup);

test('renders the JobList component', () => {
    render(<JobList/>, {wrapper: Wrapper});
});

test('matches the snapshot', async () => {
    const jobList = render(<JobList/>, {wrapper: Wrapper});
    await waitFor(() => {
        expect(JoblyApi.getAllJobs).toHaveBeenCalled();
    });
    expect(jobList).toMatchSnapshot();
});

test('displays the correct content', async () => {
    const {getByText} = render(<JobList/>, {wrapper: Wrapper});
    await waitFor(() => {
        expect(JoblyApi.getAllJobs).toHaveBeenCalled();
    })
    expect(getByText('Accommodation manager')).toBeInTheDocument();
});

test('displays loading if loading', () => {
    const {getByText} = render(<JobList/>, {wrapper: Wrapper});
    expect(getByText('Loading', {exact: false})).toBeInTheDocument();
});

test('redirects to home if not logged in', () =>{
    const {getByText} = render(
        <MemoryRouter>
            <JobList />
        </MemoryRouter>);
    expect(getByText('Redirected to /')).toBeInTheDocument();
});