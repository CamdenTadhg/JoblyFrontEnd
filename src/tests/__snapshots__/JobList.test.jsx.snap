// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="JobList-search"
        >
          <div
            class="SearchForm"
          >
            <form
              class="SearchForm-form"
            >
              <div
                class="input-item"
              >
                <label
                  for="title"
                >
                  Title
                  :
                </label>
                <input
                  id="title"
                  name="title"
                  value=""
                />
              </div>
              <div
                class="input-item"
              >
                <label
                  for="salary"
                >
                  Salary
                  :
                </label>
                <input
                  id="salary"
                  name="salary"
                  value=""
                />
              </div>
              <div
                class="input-item"
              >
                <label
                  for="equity"
                >
                  Equity
                  :
                </label>
                <input
                  id="equity"
                  name="equity"
                  type="checkbox"
                  value="false"
                />
              </div>
            </form>
          </div>
        </div>
        <div
          class="JobList-jobs"
        >
          <div
            class="JobCard-card card"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title"
              >
                <b>
                  Accommodation manager
                </b>
              </div>
              <div>
                Salary: 
                126000
              </div>
              <div>
                Equity: 
                YES
              </div>
              <button
                class="ApplyButton"
                disabled=""
              >
                Applied
              </button>
            </div>
          </div>
          <div
            class="JobCard-card card"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title"
              >
                <b>
                  Accountant, chartered certified
                </b>
              </div>
              <div>
                Salary: 
                175000
              </div>
              <div>
                Equity: 
                YES
              </div>
              <button
                class="ApplyButton"
              >
                Apply
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="JobList-search"
      >
        <div
          class="SearchForm"
        >
          <form
            class="SearchForm-form"
          >
            <div
              class="input-item"
            >
              <label
                for="title"
              >
                Title
                :
              </label>
              <input
                id="title"
                name="title"
                value=""
              />
            </div>
            <div
              class="input-item"
            >
              <label
                for="salary"
              >
                Salary
                :
              </label>
              <input
                id="salary"
                name="salary"
                value=""
              />
            </div>
            <div
              class="input-item"
            >
              <label
                for="equity"
              >
                Equity
                :
              </label>
              <input
                id="equity"
                name="equity"
                type="checkbox"
                value="false"
              />
            </div>
          </form>
        </div>
      </div>
      <div
        class="JobList-jobs"
      >
        <div
          class="JobCard-card card"
        >
          <div
            class="card-body"
          >
            <div
              class="card-title"
            >
              <b>
                Accommodation manager
              </b>
            </div>
            <div>
              Salary: 
              126000
            </div>
            <div>
              Equity: 
              YES
            </div>
            <button
              class="ApplyButton"
              disabled=""
            >
              Applied
            </button>
          </div>
        </div>
        <div
          class="JobCard-card card"
        >
          <div
            class="card-body"
          >
            <div
              class="card-title"
            >
              <b>
                Accountant, chartered certified
              </b>
            </div>
            <div>
              Salary: 
              175000
            </div>
            <div>
              Equity: 
              YES
            </div>
            <button
              class="ApplyButton"
            >
              Apply
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
